pkg/kedge/http/lbtransport/body.go

```diff
@@ -0,0 +1,62 @@
+package lbtransport
+
+import (
+	"bytes"
+	"io"
+)
+
+type lazyBufferedReader struct {
+	wrapped io.Reader
+	buf     []byte
+	offset  int
+}
+
+func (*lazyBufferedReader) Close() error {
+	return nil
+}
+
+// rewind allows lazyBufferedReader to be read again.
+func (b *lazyBufferedReader) rewind() {
+	if b == nil {
+		return
+	}
+	b.offset = 0
+}
+
+func (b *lazyBufferedReader) Read(p []byte) (n int, err error) {
+	if b == nil {
+		return 0, io.EOF
+	}
+	defer func() { b.offset += n }()
+
+	if len(b.buf)-b.offset > 0 {
+		n, err = bytes.NewReader(b.buf).Read(p)
+		if err != nil {
+			return n, err
+		}
+		if n == len(p) {
+			return n, nil
+		}
+	}
+
+	tmp := bytes.NewBuffer(b.buf)
+	n64, err := tmp.ReadFrom(io.LimitReader(b.wrapped, int64(len(p)-n)))
+	defer func() { n += int(n64) }()
+
+	if n64 == 0 {
+		return 0, io.EOF
```

Comment:
Multiple points here:
- `n64 == 0` does not mean that `n == 0` hence we might still be returning non-zero-length content that has already been read into `p` at this point.
- We might not be passing through a non-EOF error here. As [I understand it](https://golang.org/src/io/io.go?s=14885:14942#L431) the `io.LimitReader` will pass through any error from the underlying reader and while `n64` might still be set to 0.

My suggestion would be to invert the condition and move the 3 lines of code below into the `if` body. This has the advantages of:
1. Allowing to simply increment `n += int(n64)` without relying on a `defer`.
2. Relying on a single `return n, err` at the end of the function.

Reply:
Good catch!